storage/paths/db_index.go

```diff
@@ -0,0 +1,999 @@
+			}
+		}
+
+		// Acquire write locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sector_location
+							 	SET write_ts = NOW(), write_lock_owner = $1
+							 	WHERE miner_id=$2
+							 		AND sector_num=$3
+							 		AND sector_filetype = ANY($4)`,
+			lockUuid.String(),
+			sector.Miner,
+			sector.Number,
+			write.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring write locks for sector %v fails with err: %v", sector, err)
+		}
+
+		// Acquire read locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sector_location
+							 	SET read_ts = NOW(), read_refs = read_refs + 1
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner,
+			sector.Number,
+
+			read.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring read locks for sector %v fails with err: %v", sector, err)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (dbi *DBIndex) unlock(sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType, lockUuid uuid.UUID) (bool, error) {
+	ctx := context.Background()
+
+	if read|write == 0 {
+		return false, nil
+	}
+
+	if read|write > (1<<storiface.FileTypes)-1 {
+		return false, xerrors.Errorf("unknown file types specified")
+	}
+
+	// Relinquish write locks
+	_, err := dbi.harmonyDB.Exec(ctx,
+		`UPDATE sector_location
+				SET write_ts = NULL, write_lock_owner = NULL
+				WHERE miner_id=$1
+					AND sector_num=$2
+					AND write_lock_owner=$3
+					AND sector_filetype = ANY($4)`,
+		sector.Miner,
+		sector.Number,
+		lockUuid.String(),
+		write.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing write locks for sector %v fails with err: %v", sector, err)
+	}
+
+	// Relinquish read locks
+	_, err = dbi.harmonyDB.Exec(ctx,
+		`UPDATE sector_location
+				SET read_refs = read_refs-1,
+					read_ts = CASE WHEN read_refs - 1 = 0 THEN NULL ELSE read_ts END
+				WHERE miner_id=$1
+					AND sector_num=$2
+					AND sector_filetype = ANY($3)`,
+		sector.Miner,
+		sector.Number,
+		read.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing read locks for sector %v fails with err: %v", sector, err)
+	}
+
+	return true, nil
+
+}
+
+func (dbi *DBIndex) StorageLock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) error {
+
+	retries := 5
+	maxWaitTime := 300 // Set max wait time to 5 minutes
+
+	waitTime := 1
+	// generate uuid for this lock owner
+	lockUuid := uuid.New()
+
+	// retry with exponential backoff and block until lock is acquired
+	for {
+		locked, err := dbi.lock(ctx, sector, read, write, lockUuid)
+		// if err is not nil and is not because we cannot acquire lock, retry
+		if err != nil && !strings.Contains(err.Error(), "cannot acquire") {
```

Comment:
Can we check for some defined error? Define `var errFailedToAcquireLock = errors.New("...")` somewhere above, and return `xerrors.Errorf("acquiring write locks for sector..: %w", ..., errFailedToAcquireLock)`, then check here with `errors.Is` / `errors.As`.

Reply:
done